Build failure: whois (on Darwin)

### Steps To Reproduce

Steps to reproduce the behavior:
1. build `whois` on darwin on nixos-unstable

### Build log

See https://hydra.nixos.org/build/237625043/nixlog/1

Relevant bit:

```
./utils.h:62:7: error: 'malloc' attribute takes no arguments
void *MALLOC_FREE NONNULL do_nofail(void *ptr, const char *file, const int line)
      ^
./utils.h:20:37: note: expanded from macro 'MALLOC_FREE'
# define MALLOC_FREE __attribute__((malloc(free)))
```

### Additional context

Failing due to https://github.com/NixOS/nixpkgs/pull/259969

Appears to be due to [this upstream change](https://github.com/rfc1036/whois/compare/v5.5.18...v5.5.19#diff-a00a459441a1c86324aacc5bdbb53461d439d2af337849915891aa3dc281936e). Though it’s not clear to me why it appears that `__GNUC__` is defined when using clang.

### Notify maintainers

@fpletz 

### Metadata

Hydra, can replicate locally.

--- a/pkgs/tools/networking/whois/clang.patch
+++ b/pkgs/tools/networking/whois/clang.patch
@@ -1,0 +1,11 @@
+--- a/utils.h
++++ b/utils.h
+@@ -15,7 +15,7 @@
+ #endif
+ 
+ /* Portability macros */
+-#ifdef __GNUC__
++#if defined __GNUC__ && ! defined __clang__
+ # define NORETURN __attribute__((noreturn))
+ # define MALLOC_FREE __attribute__((malloc(free)))
+ # define NONNULL __attribute__((returns_nonnull))

--- a/pkgs/tools/networking/whois/default.nix
+++ b/pkgs/tools/networking/whois/default.nix
@@ -11,6 +11,10 @@ stdenv.mkDerivation rec {
     hash = "sha256-4mFn5cY7ipAU4vOiHC2s69fxYJwShQEQ1eA8t5JvOP0=";
   };
 
+  patches = [
+    ./clang.patch
+  ];
+
   nativeBuildInputs = [ perl gettext pkg-config ];
   buildInputs = [ libidn2 libiconv ];
 

